Refactor non-negative CP and provide nndsvd initialization

aarmey opened the pull request:
- tl.clip seems to be broken for the MXnet backend when only `a_min` or `a_max` are used alone.
- Perhaps it's worth re-factoring the non-negative methods into a separate file? They seem to be almost entirely separate.

JeanKossaifi commented:
Thanks @aarmey  -- will check mxnet's clip. 

> Perhaps it's worth re-factoring the non-negative methods into a separate file? They seem to be almost entirely separate.

I agree -- let's separate them!

aarmey commented:
Great. I'll make those changes here.

aarmey commented on tensorly/decomposition/_nn_cp.py:
```diff
@@ -0,0 +1,413 @@
+        # TODO: Test this
+        try:
+            kt = CPTensor(init)
+        except ValueError:
```
Should this exception be within CPTensor?

JeanKossaifi replied:
Possibly but in this case we also want to alert the user about what he did wrong (e.g. they should get an error message explaining what is wrong -- not a CP init-- and hinting at how to fix it. If the error is raised in CPTensor it might be harder for them to understand and fix the issue?

aarmey commented on tensorly/decomposition/_cp.py:
```diff
@@ -44,9 +43,7 @@ def initialize_cp(tensor, rank, init='svd', svd='numpy_svd', random_state=None,
-        # factors = [tl.tensor(rng.random_sample((tensor.shape[i], rank)), **tl.context(tensor)) for i in range(tl.ndim(tensor))]
-        # kt = CPTensor((None, factors))
-        return random_cp(tl.shape(tensor), rank, normalise_factors=False, **tl.context(tensor))
+        kt = random_cp(tl.shape(tensor), rank, normalise_factors=False, **tl.context(tensor))
```
This can't immediately return because it potentially has to be normalized later.

aarmey commented:
Addresses #203.

aarmey commented on tensorly/decomposition/_nn_cp.py:
```diff
@@ -0,0 +1,413 @@
+        if orthogonalise and iteration <= orthogonalise:
+            for i, f in enumerate(factors):
+                if min(tl.shape(f)) >= rank:
+                    factors[i] = tl.abs(tl.qr(f)[0])
```
This seems like a really bad approach to orthogonalization. An alternative might be [pivoting](http://proceedings.mlr.press/v97/zhang19r/zhang19r.pdf)? I'd suggest removing it for now...

JeanKossaifi replied:
This orthogonalization was taken from the paper "[Orthgonalized ALS for Tensor Decomposition](http://proceedings.mlr.press/v70/sharan17a/sharan17a.pdf)", Sharan and Valiant, ICML 2017.

It was added to the original CP and was never intended to really be used with the non-negative version and might be problematic indeed. Pivoting would be great to have instead! Happy to remove in the meantime.

aarmey commented:
This should be ready for review @JeanKossaifi. The only remaining test failure is from tensorflow, which doesn't seem to accept boolean indexing? Could use your advice on how to resolve that.

JeanKossaifi commented on tensorly/decomposition/_nn_cp.py:
```diff
@@ -0,0 +1,413 @@
+            idx = W < np.finfo(np.float32).eps
+
+            if nntype == "nndsvd":
+                W = tl.index_update(W, tl.index[idx], 0.0)
```
For this we can use proximal.soft_thresholding with esp as threshold.

JeanKossaifi commented on tensorly/decomposition/_nn_cp.py:
```diff
@@ -0,0 +1,413 @@
+                W = tl.index_update(W, tl.index[idx], 0.0)
+            elif nntype == "nndsvda":
+                avg = tl.mean(tensor)
+                W = tl.index_update(W, tl.index[idx], avg)
```
A less-efficient way to do this is using something like this
```python
tl.where(W < eps, tl.ones(tl.shape(scales), **tl.context(factor))*avg, W)
```
But this should work for all backends.

JeanKossaifi commented on tensorly/decomposition/_nn_cp.py:
```diff
@@ -0,0 +1,413 @@
+                H = tl.index_update(H, tl.index[j, :], lbd * v)
+
+            # After this point we no longer need H
+            idx = W < np.finfo(np.float32).eps
```
This could be just `eps = tl.eps(tensor.dtype)` and then using it as in the comments below.

JeanKossaifi commented:
Thanks for the PR @aarmey !

I'm wondering whether it would be worth having a separate function for the non-negative SVD and calling it in the CP-init as it might be useful in other places (e.g. non-negative Tucker).

aarmey commented:
Thanks @JeanKossaifi. I agree about separating nndsvd into a function so that is done here. I've applied your other suggestions, and it looks like everything is building (aside from that flakey test).

JeanKossaifi commented:
Thanks @aarmey -- will review asap. It seems there is an issue with Travis that was going through my personal credit for testing. As I ran out, testing has currently been disabled, I'll try to fix the issue or switch to Github's workflows instead.

JeanKossaifi commented:
I switched to Github actions for the tests -- any commit should trigger it.

Philosophical question: should `nnsvd` take an SVD as input and make it non-negative as currently done, or should it directly take a tensor and apply both steps (we can then have a intermediate function, e.g. `make_svd_non_negative`). What do you think?

aarmey commented:
I like having *a* function that just takes the SVD as input because it doesn't depend in any way on the manner by which the SVD was constructed. A function that offers `make_svd_non_negative` seems natural, but we don't have a `make_svd` on the other hand. Would a `make_svd`/`make_svd_non_negative` help to simplify `initialize_nn_cp`, `initialize_cp`...?

JeanKossaifi commented:
> I like having _a_ function that just takes the SVD as input because it doesn't depend in any way on the manner by which the SVD was constructed. 

I agree. 

The non-negative init is a two step process i) apply regular SVD: `U, S, V = svd(...)` and ii) (only in the non-negative case) `make_svd_non_negative(U, S, V)` -- I suggested this name to make explicit that it takes an _existing_ SVD and *makes* it non-negative (and then we could optionally have a function `nnsvd` that directly takes as input a tensor and applies both steps). The reasoning was that a user might expect any kind of SVD to take a matrix as input, but I am also happy to just keep the current naming! What do you think?

aarmey commented:
Oh! Completely agree. The function is renamed here now.

I believe I've fixed `test_parafac` now. Previously it was making a tensor filled with random numbers of undetermined rank, then decomposing with rank 6. I changed this to make sure it's using a tensor that actually has corresponding low-rank structure. I also used `pytest.parameterize` a bit to clean up the various combination of tested options.

JeanKossaifi commented:
Awesome, thanks a lot @aarmey! 
You may want to add yourself to the author list for the nn-svd? Otherwise happy to merge! :) 

I thought I had fix that pesky tucker validation issue -- will dig into it.

aarmey commented:
Done. Thanks @JeanKossaifi
